var/spack/repos/builtin/packages/bcl2fastq2/package.py

```diff
@@ -65,7 +66,9 @@ class Bcl2fastq2(Package):
     # v2.19.1.403 is only available via ftp.
     # who knows what the future will hold.
     def url_for_version(self, version):
-        if version.string == '2.19.1.403':
+        if version.string == '2.20.0.422':
+            return "ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/software/bcl2fastq/bcl2fastq2-v2-20-0-tar.zip"
+        elif version.string == '2.19.1.403':
```

Comment:
It looks like 2.18 is no longer available for download either. At this point, you can get rid of the entire `url_for_version` and add the `url` below `homepage`.

Reply:
Some folks (e.g. me at my old gig, and perhaps @peetsv) have the old tarballs cached so that we can reproduce our environment even after Illumina moves forward.

What will redoing the url bits do to such folk (I no longer have access to those tarballs and can't test it...).

Reply:
I don't think that will change anything. But I can't remember how the tarball is named in the mirror.

Reply:
Do you still want this fix or not?

Reply:
I'd say clean it up.  If someone who has an old tarball stashed runs into trouble it'll be easy enough to fix, but I have no way to test it, and now that the tarball's gone there's no way to test the old ones either.

Reply:
But if we remove `url_for_version` and use, for example, `url = "ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/software/bcl2fastq/bcl2fastq2-v2.19.1-tar.zip"`, how is it gonna find `bcl2fastq2-v2-20-0-tar.zip`, which has hyphens instead of dots, from the specified versions?

Reply:
Or should I specify the correct `url` for each version?

Reply:
such as:

`version('2.19.1', 'baba7a02767fd868e87cb36781d2be26', url='ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/software/bcl2fastq/bcl2fastq2-v2.19.1-tar.zip')`

`version('2.20.0', '4dc99f1af208498b7279b66556329488', url='ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/software/bcl2fastq/bcl2fastq2-v2-20-0-tar.zip')`

Reply:
Ah, good point. You'll still need a `url_for_version` in that case, but it will become much simpler.

Reply:
Hi! I'm a security researcher @TruffleSecurity. I found a valid API key leaked in a comment in this Issue. I highly recommend rotating the API key immediately (not just deleting the comment). Here's a blog post I wrote on remediating these types of vulnerabilities: https://trufflesecurity.com/blog/remediate-leaked-api-keys-with-key-rotation.

If you can't find the leaked API key, you can use the following [TruffleHog](https://github.com/trufflesecurity/trufflehog) command to automatically scan for it:
`trufflehog github --repo https://github.com/<your_username>/<your_reponame> --issue-comments --only-verified`

If it's not clear where/what the security issue is, please let me know. I'd be happy to help. Please note that the API key might be in the past edit of a comment.

(cc @matteoguglielmi )